msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AcquireTokenByManagedIdentitySupplier.java

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.aad.msal4j;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class AcquireTokenByManagedIdentitySupplier extends AuthenticationResultSupplier {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AcquireTokenByManagedIdentitySupplier.class);
+
+    private ManagedIdentityParameters managedIdentityParameters;
+
+    AcquireTokenByManagedIdentitySupplier(ManagedIdentityApplication managedIdentityApplication, MsalRequest msalRequest) {
+        super(managedIdentityApplication, msalRequest);
+        this.managedIdentityParameters = (ManagedIdentityParameters) msalRequest.requestContext().apiParameters();
+    }
+
+    @Override
+    AuthenticationResult execute() throws Exception {
+
+        if (StringHelper.isNullOrBlank(managedIdentityParameters.resource)) {
+            throw new MsalClientException(
+                    MsalError.SCOPES_REQUIRED,
+                    MsalErrorMessage.SCOPES_REQUIRED);
+        }
+
+        TokenRequestExecutor tokenRequestExecutor = new TokenRequestExecutor(
+                clientApplication.authenticationAuthority,
+                msalRequest,
+                clientApplication.getServiceBundle()
+        );
+
+        if (!managedIdentityParameters.forceRefresh) {
+            LOG.debug("ForceRefresh set to false. Attempting cache lookup");
+
+            try {
+                Set<String> scopes = new HashSet<>();
+                scopes.add(this.managedIdentityParameters.resource);
+                SilentParameters parameters = SilentParameters
+                        .builder(scopes)
+                        .build();
+
+                RequestContext context = new RequestContext(
+                        this.clientApplication,
+                        PublicApi.ACQUIRE_TOKEN_SILENTLY,
+                        parameters);
+
+                SilentRequest silentRequest = new SilentRequest(
+                        parameters,
+                        this.clientApplication,
+                        context,
+                        null);
+
+                AcquireTokenSilentSupplier supplier = new AcquireTokenSilentSupplier(
+                        this.clientApplication,
+                        silentRequest);
+
+                return supplier.execute();
+            } catch (MsalClientException ex) {
```

Comment:
This is surprising. Exception should not control the programming flow. 

https://softwareengineering.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why

If this is a common pattern in MSAL 4J, at least there should be some filtering on ex.Code ?

Reply:
This is how the flow is in client credential. I can add a filter on error code [cache_miss](https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/eb36681986d3ae2432e438949efdb9086476e1c4/msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AuthenticationErrorCode.java#L59)

Reply:
Ok, feel like there's already enough refactoring as part of this feature.

Reply:
Changed for managed identity. Not updating for client credentials flow in this PR.